fieldsolver/gridGlue.cpp

```diff
@@ -179,6 +189,244 @@ void feedMomentsIntoFsGrid(dccrg::Dccrg<SpatialCell,dccrg::Cartesian_Geometry>&
 
   MPI_Waitall(sendRequests.size(), sendRequests.data(), MPI_STATUSES_IGNORE);
 
+  // ---------------------------------------------------------------------------------
+  // Boxcar Smoothening
+  const int *mntDims= &momentsGrid.getLocalSize()[0];
+  std::array<Real,fsgrids::moments::N_MOMENTS> *Grid;
+  std::array<Real,fsgrids::moments::N_MOMENTS> *GridL1;
+  std::array<Real,fsgrids::moments::N_MOMENTS> *GridL2;
+  std::array<Real,fsgrids::moments::N_MOMENTS> *GridR1;
+  std::array<Real,fsgrids::moments::N_MOMENTS> *GridR2;
+  std::array<Real,fsgrids::moments::N_MOMENTS> *swap;
+  auto swapGrid= momentsGrid;
+ 
+
+  /*Kernel Weights are fractional numbers 
+
+              std::array weights[i]
+    weight= ---------------------------
+                2*halfStencilWidht+1
+
+  and are calculated as shown above while iterating over each dimensions
+  The numerator is calculated filling the weights array
+  and the denominator depends only on the stencil width. 
+  */
+  Real weightC, weightL1, weightL2, weightR1, weightR2;
+  int maxRefLevel = mpiGrid.mapping.get_maximum_refinement_level();
+  int minRefLevel = 0;
+  const int maxStencilWidth=5;
+  int blurPasses = 3;
+  int refLevel;
+  int halfStencilWidth;
+  std::array<int, maxStencilWidth> weights;
+  weights.fill(0);
+
+
+  for (int blurPass = 1; blurPass <= blurPasses; blurPass++)
+  {
+
+    // X Dimension Pass
+    #pragma omp parallel for collapse(3)
+    for (int kk = 0; kk < mntDims[2]; kk++)
+    {
+      for (int jj = 0; jj < mntDims[1]; jj++)
+      {
+        for (int ii = 0; ii < mntDims[0]; ii++)
+        {
+
+          // Dont filter if in max resolution region
+          refLevel = technicalGrid.get(ii, jj, kk)->RefLevel;
+          if (refLevel == maxRefLevel)
+          {
+                        continue;
+          }
+
+          // Set 5 point stencil for a set refinement level and below
+          if (refLevel<=minRefLevel+1)
+            {
+              halfStencilWidth=2;              
+            }
+          else{
+              halfStencilWidth=1;
+          } 
+
+          // Calculate Numerator
+          for (int index = 0; index < halfStencilWidth; index++)
+          {
+            weights.at(2 + index) = 1;
+            weights.at(2 - index) = 1;
+          }
+          // Calculate Denominator
+          int weightDenominator = 2 * halfStencilWidth + 1;
+
+          // Calculate Weights
+          weightL2 = weights.at(0) / weightDenominator;
+          weightL1 = weights.at(1) / weightDenominator;
+          weightC  = weights.at(2) / weightDenominator;
+          weightR1 = weights.at(3) / weightDenominator;
+          weightR2 = weights.at(4) / weightDenominator;
```

Comment:
Since we only have a small number (4) of possible refinement levels here, the weight values only have 4 different outcomes, right?
Instead of recalculating them for every cell, we could also prebuild the 4 possible weights arrays before the loop and then just look them up.
(This is just a performance enhancement, let's get it to work in principle fist)

Reply:
Yes I thought of that too but skipped it till now.  It is actually easier as the weight values only depend on the stencil width. So since we are using a 3 and 5 point stencil the weights have 2 possible outcomes. Will leave this for a bit later though. Trying to merge the loops first